### ISSUE TITLE
[SOL-43235] SCSt. Binder Issue with Custom Error Handling Runtime Exceptions

### ISSUE THREAD

**Issue by Mrc0113** [SolaceProducts/solace-spring-cloud#36]
The Solace Spring Cloud Stream Binder does not properly handle the situation where a custom error handler using @ServiceActivator on the global "errorChannel" or a binding specific error channel like "myDestination.myGroup.errors" throws a RuntimeException. The binder just logs the error and does not ACK or REJECT the message. This will result in either memory running out or the app not receiving anymore messages when it hits the max unacked messages per flow limit. 

Example code to reproduce: 
```java
	@Bean
	public Function<String, String> processDropoffRideAverages() {
		return tsup -> {
			System.out.println("Received tsup:" + tsup);
			throw new RuntimeException("oh no!");
		};
	}

	@ServiceActivator(inputChannel = "test/taxinyc/RideDropoffProcessorQueue.foo.errors")
	private void singleChannelErrorHandler(ErrorMessage errMsg) {
		System.out.println("CHANNEL SPECIFIC error channel:" + errMsg);
		throw new RuntimeException("Exception from channel specific error handler!");
	}
```

Example exception: 
```
2020-11-25 16:50:03.983  WARN 89032 --- [pool-2-thread-1] c.s.s.c.s.b.i.InboundXMLMessageListener  : Exception received while consuming messages from destination test/taxinyc/RideDropoffProcessorQueue.foo

org.springframework.messaging.MessageHandlingException: error occurred during processing message in 'MethodInvokingMessageProcessor' [org.springframework.integration.handler.MethodInvokingMessageProcessor@2cf70176]; nested exception is java.lang.RuntimeException: Exception from channel specific error handler!
	at org.springframework.integration.support.utils.IntegrationUtils.wrapInHandlingExceptionIfNecessary(IntegrationUtils.java:192) ~[spring-integration-core-5.3.2.RELEASE.jar:5.3.2.RELEASE]
	at org.springframework.integration.handler.MethodInvokingMessageProcessor.processMessage(MethodInvokingMessageProcessor.java:111) ~[spring-integration-core-5.3.2.RELEASE.jar:5.3.2.RELEASE]
	at org.springframework.integration.handler.ServiceActivatingHandler.handleRequestMessage(ServiceActivatingHandler.java:104) ~[spring-integration-core-5.3.2.RELEASE.jar:5.3.2.RELEASE]
	at org.springframework.integration.handler.AbstractReplyProducingMessageHandler.handleMessageInternal(AbstractReplyProducingMessageHandler.java:134) ~[spring-integration-core-5.3.2.RELEASE.jar:5.3.2.RELEASE]
	at org.springframework.integration.handler.AbstractMessageHandler.handleMessage(AbstractMessageHandler.java:62) ~[spring-integration-core-5.3.2.RELEASE.jar:5.3.2.RELEASE]
	at org.springframework.integration.dispatcher.AbstractDispatcher.tryOptimizedDispatch(AbstractDispatcher.java:115) ~[spring-integration-core-5.3.2.RELEASE.jar:5.3.2.RELEASE]
	at org.springframework.integration.dispatcher.UnicastingDispatcher.doDispatch(UnicastingDispatcher.java:133) ~[spring-integration-core-5.3.2.RELEASE.jar:5.3.2.RELEASE]
	at org.springframework.integration.dispatcher.UnicastingDispatcher.dispatch(UnicastingDispatcher.java:106) ~[spring-integration-core-5.3.2.RELEASE.jar:5.3.2.RELEASE]
	at org.springframework.integration.channel.AbstractSubscribableChannel.doSend(AbstractSubscribableChannel.java:72) ~[spring-integration-core-5.3.2.RELEASE.jar:5.3.2.RELEASE]
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:570) ~[spring-integration-core-5.3.2.RELEASE.jar:5.3.2.RELEASE]
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:520) ~[spring-integration-core-5.3.2.RELEASE.jar:5.3.2.RELEASE]
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:187) ~[spring-messaging-5.2.8.RELEASE.jar:5.2.8.RELEASE]
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:166) ~[spring-messaging-5.2.8.RELEASE.jar:5.2.8.RELEASE]
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:47) ~[spring-messaging-5.2.8.RELEASE.jar:5.2.8.RELEASE]
	at org.springframework.messaging.core.AbstractMessageSendingTemplate.send(AbstractMessageSendingTemplate.java:109) ~[spring-messaging-5.2.8.RELEASE.jar:5.2.8.RELEASE]
	at org.springframework.messaging.core.AbstractMessageSendingTemplate.send(AbstractMessageSendingTemplate.java:99) ~[spring-messaging-5.2.8.RELEASE.jar:5.2.8.RELEASE]
	at org.springframework.integration.core.ErrorMessagePublisher.publish(ErrorMessagePublisher.java:168) ~[spring-integration-core-5.3.2.RELEASE.jar:5.3.2.RELEASE]
	at org.springframework.integration.handler.advice.ErrorMessageSendingRecoverer.recover(ErrorMessageSendingRecoverer.java:83) ~[spring-integration-core-5.3.2.RELEASE.jar:5.3.2.RELEASE]
	at com.solace.spring.cloud.stream.binder.inbound.RetryableInboundXMLMessageListener.lambda$handleMessage$11(RetryableInboundXMLMessageListener.java:50) ~[spring-cloud-stream-binder-solace-core-2.1.1.jar:na]
	at org.springframework.retry.support.RetryTemplate.handleRetryExhausted(RetryTemplate.java:512) ~[spring-retry-1.2.5.RELEASE.jar:na]
	at org.springframework.retry.support.RetryTemplate.doExecute(RetryTemplate.java:351) ~[spring-retry-1.2.5.RELEASE.jar:na]
	at org.springframework.retry.support.RetryTemplate.execute(RetryTemplate.java:180) ~[spring-retry-1.2.5.RELEASE.jar:na]
	at com.solace.spring.cloud.stream.binder.inbound.RetryableInboundXMLMessageListener.handleMessage(RetryableInboundXMLMessageListener.java:44) ~[spring-cloud-stream-binder-solace-core-2.1.1.jar:na]
	at com.solace.spring.cloud.stream.binder.inbound.InboundXMLMessageListener.receive(InboundXMLMessageListener.java:115) ~[spring-cloud-stream-binder-solace-core-2.1.1.jar:na]
	at com.solace.spring.cloud.stream.binder.inbound.InboundXMLMessageListener.run(InboundXMLMessageListener.java:75) ~[spring-cloud-stream-binder-solace-core-2.1.1.jar:na]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_201]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_201]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_201]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_201]
	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_201]
Caused by: java.lang.RuntimeException: Exception from channel specific error handler!
	at org.taxi.nyc.Application.singleChannelErrorHandler(Application.java:92) ~[classes/:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_201]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_201]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_201]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_201]
	at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:171) ~[spring-messaging-5.2.8.RELEASE.jar:5.2.8.RELEASE]
	at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.invoke(InvocableHandlerMethod.java:120) ~[spring-messaging-5.2.8.RELEASE.jar:5.2.8.RELEASE]
	at org.springframework.integration.handler.support.MessagingMethodInvokerHelper$HandlerMethod.invoke(MessagingMethodInvokerHelper.java:1104) ~[spring-integration-core-5.3.2.RELEASE.jar:5.3.2.RELEASE]
	at org.springframework.integration.handler.support.MessagingMethodInvokerHelper.invokeHandlerMethod(MessagingMethodInvokerHelper.java:582) ~[spring-integration-core-5.3.2.RELEASE.jar:5.3.2.RELEASE]
	at org.springframework.integration.handler.support.MessagingMethodInvokerHelper.processInternal(MessagingMethodInvokerHelper.java:477) ~[spring-integration-core-5.3.2.RELEASE.jar:5.3.2.RELEASE]
	at org.springframework.integration.handler.support.MessagingMethodInvokerHelper.process(MessagingMethodInvokerHelper.java:355) ~[spring-integration-core-5.3.2.RELEASE.jar:5.3.2.RELEASE]
	at org.springframework.integration.handler.MethodInvokingMessageProcessor.processMessage(MethodInvokingMessageProcessor.java:108) ~[spring-integration-core-5.3.2.RELEASE.jar:5.3.2.RELEASE]
	... 28 common frames omitted
```

**Comment by Nephery** [comment]
I think there's two cases to consider here: when `max-retries == 1` and `max-retries > 1`. What you've shown is the case when `max-retries > 1`. While in the `max-retries == 1`, the flow completely shuts down which is a problem.

So going back to our flowchart draft for second:
![image](https://user-images.githubusercontent.com/11084623/100380730-65fc9380-2fe5-11eb-8d98-2fb9b954ff25.png)

We actually have control over all the exit paths from `Custom Global errorChannel Listener(s)` and `Channel Specific Error Handler`, with exception for the `SUCCESSFUL EXIT` path of `Custom Global errorChannel Listener(s)` which leads to our [SolaceErrorMessageHandler](https://github.com/SolaceProducts/solace-spring-cloud/blob/1.1.1/solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/util/SolaceErrorMessageHandler.java#L42) (this is fine).

<hr>

Going into a bit of a side tangent here...

As for the `SUCCESSFUL EXIT` path from `Channel Specific Error Handler`, we currently directly execute an `ACCEPT` auto-ack on the message (actually a `REJECT` in 2.X, but `REJECT == ACCEPT`; will be changed to properly execute an `ACCEPT` in 3.0.0). Though its not in the snippet I posted above, this is illustrated in flowchart.

Up until now, I thought the framework was the one that was acknowledging the message for this path, but it turns out, we were the ones doing it the whole time. Either way, I think this is fine since a successful exit from the `Channel Specific Error Handler` would indicate a successful handling of the error. But @Mrc0113 , in case you disagree, let me know and we can change this for 3.0.0.

For reference here are the lines of code for `max-retries == 1` and `max-retries > 1` which invoke the post-error auto-ack:
* [InboundXMLMessageListener.java#L160](https://github.com/SolaceProducts/solace-spring-cloud/blob/1.1.1/solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/inbound/InboundXMLMessageListener.java#L160)
* [RetryableInboundXMLMessageListener.java#L51](https://github.com/SolaceProducts/solace-spring-cloud/blob/1.1.1/solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/inbound/RetryableInboundXMLMessageListener.java#L51)

**On a somewhat unrelated note:** The auto-ack also executes after the [SolaceErrorMessageHandler](https://github.com/SolaceProducts/solace-spring-cloud/blob/1.1.1/solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/util/SolaceErrorMessageHandler.java#L42) executes, but is effectively a no-op in that scenario.

<hr>

As for the actual issue, what we need to do now is figure out what to do when either of these error handlers throw an exception. And for this there are two things to consider:

1. What should we do if our error handling itself throws an exception?
2. Should we explictly handle `RequeueCurrentMessageException`?

What I'm thinking we should do is that we should infinite loop a `REJECT` acknowledge upon receiving **any** error handler exceptions. This would resolve number 1, as well as ensure that the failed message is **always** acknowledged in some way.

Then if we need to handle number 2; we would do the same thing but with a `REQUEUE` acknowledgment instead of a `REJECT`.

The only caveat with this solution is that message processing for this consumer would be blocked until the problematic message is successfully processed.

**Comment by Nephery** [comment]
Discussed with @Mrc0113 :

* We should change default max-redelivery of 3 for regular consumer group queue and error queue.
> What I'm thinking we should do is that we should infinite loop a REJECT acknowledge upon receiving any error handler exceptions. This would resolve number 1, as well as ensure that the failed message is always acknowledged in some way.
* Add a max retries for error handling retries instead of infinite loop, in which case we should fallback requeue, log and discard, or crash the app?
* Should be smart to process all messages before flow rebind? e.g. if multiple messages are inflight from the same flow, what do we do if one requeued (i.e. flow rebind), and the other is accepted/rejected/requeued?

**Comment by Nephery** [comment]
> Should we change the default max-redelivery to 3 for regular consumer group queues and error queues? The current default is infinite.

The default will remain at infinite. We should provide guidance in the documentation that users should either enable `autoBindErrorQueue` or create a DMQ and configure a TTL/max-redelivery.

> Instead of doing an infinite loop for acknowledgment exceptions, should we add a max-attempt consumer config option for error handling retries? In which case, once the max number of attempts have been exhausted, should we fallback to REQUEUE (i.e. rebind the flow), log and discard the message, or crash the app?

There's really only one situation that we need to directly address for acknowledgment exceptions; when republishing messages to error queues. In which case, we should retry 3 times, then fallback to doing a REQUEUE. The number of attempts will be configurable under a new consumer config option named `errorQueuePublishMaxAttempts`.

Also note that error queue publishing should be a blocking action. i.e. we must wait for publish response.

> Should the binder be smart enough to process all messages before flow rebind? e.g. if multiple messages are inflight from the same flow and one message were to be requeued (i.e. flow rebind), should the binder delay the rebind until the other messages are accepted/rejected/requeued? Otherwise, the binder will have to throw an exception if any other message were to attempt acknowledge after a rebind.

FYI this is only an issue when using asynchronous client acknowledgments. Here was the proposed solution for this:

When REQUEUE happens, stop the flow, stop the consumer receives, and wait 10 seconds (configurable) for all received messages to be ack'd. If either all the outstanding messages are acknowledged or the timeout expires, then proceed with the rebind. In the event of a timeout expiry, any outstanding messages will be considered obsolete and will throw an exception if acknowledged.

We want to both stop the flow and the consumer receives both because stopping the flow is an asynchronous action and because there may already be messages cached by the API. We do not want to receive **any** messages during this wait time since that introduces a race condition between the dispatch of cached messages and the detection of when all messages are acknowledged.

The message acknowledgment wait time will configurable under a consumer config option named `flowPreRebindWaitTimeout`.

> Should we explictly handle RequeueCurrentMessageException? Is this something we should even support for asynchronous consumers? Because the documentation and the class itself only specifices it for polled consumers.

Unless someone says otherwise, we'll proceed with what we proposed to support `RequeueCurrentMessageException`. i.e. do a REQUEUE instead of a REJECT if this exception is thrown from a error message handler.

@Mrc0113 did I miss anything?

**Comment by Mrc0113** [comment]
Hey @Nephery, nope I think you covered it all. Thanks!

> In the event of a timeout expiry, any outstanding messages will be considered obsolete and will throw an exception if acknowledged.
This may just need to be an error logged.  I don't know if throwing an exception will work in this scenario as calling an ACK isn't synchronous.

**Comment by Nephery** [comment]
> This may just need to be an error logged. I don't know if throwing an exception will work in this scenario as calling an ACK isn't synchronous.

As per [XMLMessage's javadoc](https://docs.solace.com/API-Developer-Online-Ref-Documentation/java/com/solacesystems/jcsmp/XMLMessage.html#ackMessage()), an exception is thrown when the message's consumer is closed (which is the case after a rebind).

Personally, I'm pretty indifferent regarding how to handle this scenario, so I can change that to just a warning log if you want.

**Comment by Nephery** [comment]
Closed with #75
